File: qrs-core-python/src/qrs/services/revocationService.py
"""RevocationService: revocation, blocking, and lifecycle checks.

Revocation works through signed statements (offline-friendly):

- ``revokeTcert`` targets a TCert (prospective/retrospective) or a key (target
  kind ``key``, which invalidates all TCerts of that key).
- ``blockSdoc`` / ``unblockSdoc`` target a single SDoc by its id.

Checks distinguish prospective vs retrospective, and always evaluate key
revocation before TCert revocation.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from ..crypto.providers import KeyPairMaterial
from ..deps import ServiceDeps
from ..envelope import parse_signed_object, split_tcert_id, tcert_id_of
from ..errors import QrsAuthorizationError, QrsCryptoError, QrsNotFoundError
from ..id import to_hex
from .statement import StatementOptions, StatementTarget, build_statement, verify_statement
from ..storage.interfaces import RevocationEntry

__all__ = [
    "RevocationService",
    "RevocationCheck",
    "RevokeTcertParams",
    "RevokeKeyParams",
    "BlockSdocParams",
    "StatementResult",
]


@dataclass
class RevocationCheck:
    state: str  # 'valid' | 'invalid'
    message: str | None = None


@dataclass
class RevokeTcertParams:
    signer_key_id: str
    target_tcert_id: str
    type: str  # 'prospective' | 'retrospective'
    reason: str | None = None
    issued_at: int | None = None


@dataclass
class RevokeKeyParams:
    signer_key_id: str
    target_key_id: str
    reason: str | None = None
    issued_at: int | None = None


@dataclass
class BlockSdocParams:
    signer_key_id: str
    target_sdoc_id: str
    reason: str | None = None
    issued_at: int | None = None


@dataclass
class StatementResult:
    statement_id: str
    bytes: bytes


class RevocationService:
    def __init__(self, deps: ServiceDeps) -> None:
        self._deps = deps

    async def revoke_tcert(self, params: RevokeTcertParams) -> StatementResult:
        target = split_tcert_id(params.target_tcert_id)
        authorized = params.signer_key_id == target[0] or await self.is_authorized_ca(
            params.signer_key_id, params.target_tcert_id
        )
        if not authorized:
            raise QrsAuthorizationError("Signer is not authorized to revoke this TCert")
        issued_at = params.issued_at if params.issued_at is not None else self._deps.clock.now()
        built = await self._sign_statement(
            params.signer_key_id,
            "revokeTcert",
            StatementTarget(kind="tcert", key_id=target[0], certificate_number=target[1]),
            issued_at,
            StatementOptions(reason=params.reason, revocation_type=params.type),
        )
        await self._deps.revocation_store.add_revoked_tcert(
            params.target_tcert_id,
            RevocationEntry(type=params.type, issued_at=issued_at, reason=params.reason),
        )
        return built

    async def revoke_key(self, params: RevokeKeyParams) -> StatementResult:
        if params.signer_key_id != params.target_key_id:
            raise QrsAuthorizationError("Only the key owner may revoke its own key")
        issued_at = params.issued_at if params.issued_at is not None else self._deps.clock.now()
        built = await self._sign_statement(
            params.signer_key_id,
            "revokeTcert",
            StatementTarget(kind="key", key_id=params.target_key_id),
            issued_at,
            StatementOptions(reason=params.reason, revocation_type="retrospective"),
        )
        await self._deps.revocation_store.add_revoked_key(
            params.target_key_id,
            RevocationEntry(type="retrospective", issued_at=issued_at, reason=params.reason),
        )
        return built

    async def block_sdoc(self, params: BlockSdocParams) -> StatementResult:
        sdoc_bytes = await self._deps.document_store.get(params.target_sdoc_id)
        if not sdoc_bytes:
            raise QrsNotFoundError(f"SDoc not found: {params.target_sdoc_id}")
        parsed = parse_signed_object(sdoc_bytes)
        tcert_key_id = to_hex(parsed.data["tcertKeyId"])
        tcert_id = tcert_id_of(tcert_key_id, parsed.data["tcertNumber"])
        authorized = params.signer_key_id == tcert_key_id or await self.is_authorized_ca(
            params.signer_key_id, tcert_id
        )
        if not authorized:
            raise QrsAuthorizationError("Signer is not authorized to block this SDoc")
        issued_at = params.issued_at if params.issued_at is not None else self._deps.clock.now()
        built = await self._sign_statement(
            params.signer_key_id,
            "blockSdoc",
            StatementTarget(kind="sdoc", sdoc_id=params.target_sdoc_id),
            issued_at,
            StatementOptions(reason=params.reason),
        )
        await self._deps.revocation_store.add_blocked_sdoc(
            params.target_sdoc_id, {"issued_at": issued_at, "reason": params.reason}
        )
        return built

    async def unblock_sdoc(self, params: BlockSdocParams) -> StatementResult:
        sdoc_bytes = await self._deps.document_store.get(params.target_sdoc_id)
        if not sdoc_bytes:
            raise QrsNotFoundError(f"SDoc not found: {params.target_sdoc_id}")
        parsed = parse_signed_object(sdoc_bytes)
        tcert_key_id = to_hex(parsed.data["tcertKeyId"])
        tcert_id = tcert_id_of(tcert_key_id, parsed.data["tcertNumber"])
        authorized = params.signer_key_id == tcert_key_id or await self.is_authorized_ca(
            params.signer_key_id, tcert_id
        )
        if not authorized:
            raise QrsAuthorizationError("Signer is not authorized to unblock this SDoc")
        issued_at = params.issued_at if params.issued_at is not None else self._deps.clock.now()
        built = await self._sign_statement(
            params.signer_key_id,
            "unblockSdoc",
            StatementTarget(kind="sdoc", sdoc_id=params.target_sdoc_id),
            issued_at,
            StatementOptions(reason=params.reason),
        )
        await self._deps.revocation_store.remove_blocked_sdoc(params.target_sdoc_id)
        return built

    async def check_revocation(self, tcert_id: str, key_id: str, issued_at: int, sdoc_id: str) -> RevocationCheck:
        key_entry = await self._deps.revocation_store.get_revoked_key(key_id)
        if key_entry:
            return RevocationCheck(
                state="invalid",
                message=f"issuer key revoked ({key_entry.reason or 'key compromised'})",
            )
        tcert_entry = await self._deps.revocation_store.get_revoked_tcert(tcert_id)
        if tcert_entry:
            if tcert_entry.type == "retrospective":
                return RevocationCheck(state="invalid", message="TCert retrospectively revoked")
            if issued_at >= tcert_entry.issued_at:
                return RevocationCheck(
                    state="invalid",
                    message="TCert prospectively revoked for documents issued at/after revocation",
                )
            return RevocationCheck(state="valid", message="document issued before prospective revocation")
        block = await self._deps.revocation_store.get_blocked_sdoc(sdoc_id)
        if block:
            return RevocationCheck(state="invalid", message="SDoc is blocked")
        return RevocationCheck(state="valid")

    async def is_authorized_ca(self, signer_key_id: str, target_tcert_id: str) -> bool:
        attestations = await self._deps.trust_store.get_attestations(target_tcert_id)
        for att in attestations:
            if att.ca_key_id != signer_key_id:
                continue
            ca_bytes = await self._deps.certificate_store.get(att.ca_tcert_id)
            if not ca_bytes:
                continue
            try:
                parse_signed_object(ca_bytes)
            except Exception:
                continue
            if await self._deps.revocation_store.get_revoked_tcert(att.ca_tcert_id):
                continue
            if await self._deps.revocation_store.get_revoked_key(signer_key_id):
                continue
            if await self._deps.trust_store.is_distrusted(att.ca_tcert_id):
                continue
            if not await self._deps.trust_store.is_pinned(att.ca_tcert_id) and not await self._deps.trust_store.is_ca(
                att.ca_tcert_id
            ):
                continue
            return True
        return False

    async def _sign_statement(
        self,
        signer_key_id: str,
        action: str,
        target: StatementTarget,
        issued_at: int,
        options: StatementOptions,
    ) -> StatementResult:
        priv = await self._deps.private_key_store.load(signer_key_id)
        if not priv:
            raise QrsNotFoundError(f"Signer private key not available: {signer_key_id}")
        pub = await self._deps.public_key_store.load(signer_key_id)
        if not pub:
            raise QrsNotFoundError(f"Signer public key not found: {signer_key_id}")
        provider = self._deps.crypto_registry.get(pub["algorithm"])
        key_pair = KeyPairMaterial(
            algorithm=pub["algorithm"],
            public_jwk=pub["public_jwk"],
            private_jwk=priv["private_jwk"],
        )
        built = build_statement(action, target, issued_at, options, key_pair, provider)
        if not verify_statement(built.parsed, pub["public_jwk"], provider):
            raise QrsCryptoError("Statement failed signature verification")
        return StatementResult(statement_id=built.statement_id, bytes=built.bytes)

File: src/qrs-core-js/src/services/revocationService.ts
/**
 * RevocationService: revocation, blocking, and lifecycle checks.
 *
 * Revocation works through signed statements (offline-friendly):
 *   - `revokeTcert` targets a TCert (prospective/retrospective) or a key (target
 *     kind `key`, which invalidates all TCerts of that key).
 *   - `blockSdoc` / `unblockSdoc` target a single SDoc by its id.
 *
 * Checks distinguish prospective vs retrospective, and always evaluate key
 * revocation before TCert revocation.
 */
import type { PublicJwk } from '../crypto/jwk.js';
import { QrsAuthorizationError, QrsCryptoError, QrsNotFoundError } from '../errors.js';
import { splitTcertId, tcertIdOf } from '../signedObject/signedObject.js';
import { parseSignedObject, tcertNumberOf } from '../signedObject/signedObject.js';
import type { KeyId, RevocationType, SdocId, StatementId, TcertId } from '../types.js';
import type { ServiceDeps } from './deps.js';
import { buildStatement, verifyStatement } from './statement.js';
import type { ComponentState } from './trustService.js';

export interface RevocationCheck {
  state: ComponentState;
  message?: string;
}

export interface RevokeTcertParams {
  signerKeyId: KeyId;
  targetTcertId: TcertId;
  type: RevocationType;
  reason?: string;
  issuedAt?: number;
}

export interface RevokeKeyParams {
  signerKeyId: KeyId;
  targetKeyId: KeyId;
  reason?: string;
  issuedAt?: number;
}
export interface RevokeAttestationParams {
  caTcertId: TcertId;
  targetTcertId: TcertId;
  reason?: string;
  issuedAt?: number;
}

export interface BlockSdocParams {
  signerKeyId: KeyId;
  targetSdocId: SdocId;
  reason?: string;
  issuedAt?: number;
}

export interface StatementResult {
  statementId: StatementId;
  bytes: Uint8Array;
}

export class RevocationService {
  constructor(private readonly deps: ServiceDeps) {}

  /** Revoke this CA's attestation without revoking the target TCert itself. */
  async revokeAttestation(params: RevokeAttestationParams): Promise<StatementResult> {
    const ca = splitTcertId(params.caTcertId);
    const target = splitTcertId(params.targetTcertId);
    const attestations = await this.deps.trustStore.getAttestations(params.targetTcertId);
    if (!attestations.some((att) => att.caTcertId === params.caTcertId)) {
      throw new QrsAuthorizationError('This CA has not attested the target TCert');
    }
    const issuedAt = params.issuedAt ?? this.deps.clock.now();
    const built = await this.signStatement(
      ca.keyId,
      'revokeAttestation',
      { kind: 'tcert', keyId: target.keyId, certificateNumber: target.certificateNumber },
      issuedAt,
      { reason: params.reason, revocationType: 'prospective' }
    );
    await this.deps.revocationStore.addRevokedAttestation(params.targetTcertId, params.caTcertId, {
      type: 'prospective', issuedAt, reason: params.reason, byKeyId: ca.keyId,
      byTcertId: params.caTcertId, statementBytes: built.bytes,
    });
    return built;
  }

  /** Revoke a TCert (prospective or retrospective). */
  async revokeTcert(params: RevokeTcertParams): Promise<StatementResult> {
    const target = splitTcertId(params.targetTcertId);
    const authorized =
      params.signerKeyId === target.keyId ||
      (await this.isAuthorizedCa(params.signerKeyId, params.targetTcertId));
    if (!authorized) {
      throw new QrsAuthorizationError('Signer is not authorized to revoke this TCert');
    }
    const issuedAt = params.issuedAt ?? this.deps.clock.now();
    const built = await this.signStatement(
      params.signerKeyId,
      'revokeTcert',
      { kind: 'tcert', keyId: target.keyId, certificateNumber: target.certificateNumber },
      issuedAt,
      { reason: params.reason, revocationType: params.type }
    );
    await this.deps.revocationStore.addRevokedTcert(params.targetTcertId, {
      type: params.type,
      issuedAt,
      reason: params.reason,
      // Attribute the revocation to the signing CA/issuer so per-CA revocation
      // (one CA revokes, another still trusts) is representable and transparent.
      byKeyId: params.signerKeyId,
      statementBytes: built.bytes,
    });
    return built;
  }

  /** Revoke a key: invalidates all TCerts derived from it. */
  async revokeKey(params: RevokeKeyParams): Promise<StatementResult> {
    if (params.signerKeyId !== params.targetKeyId) {
      throw new QrsAuthorizationError('Only the key owner may revoke its own key');
    }
    const issuedAt = params.issuedAt ?? this.deps.clock.now();
    const built = await this.signStatement(
      params.signerKeyId,
      'revokeTcert',
      { kind: 'key', keyId: params.targetKeyId },
      issuedAt,
      { reason: params.reason, revocationType: 'retrospective' }
    );
    await this.deps.revocationStore.addRevokedKey(params.targetKeyId, {
      type: 'retrospective',
      issuedAt,
      reason: params.reason,
      byKeyId: params.signerKeyId,
      statementBytes: built.bytes,
    });
    return built;
  }

  /** Block one individual SDoc. */
  async blockSdoc(params: BlockSdocParams): Promise<StatementResult> {
    const sdocBytes = await this.deps.documentStore.get(params.targetSdocId);
    if (!sdocBytes) throw new QrsNotFoundError(`SDoc not found: ${params.targetSdocId}`);
    const parsed = parseSignedObject(sdocBytes);
    const tcertKeyId = parsed.signerKeyId;
    const tcertId = tcertIdOf(tcertKeyId, tcertNumberOf(parsed));
    const authorized =
      params.signerKeyId === tcertKeyId || (await this.isAuthorizedCa(params.signerKeyId, tcertId));
    if (!authorized) {
      throw new QrsAuthorizationError('Signer is not authorized to block this SDoc');
    }
    const issuedAt = params.issuedAt ?? this.deps.clock.now();
    const built = await this.signStatement(
      params.signerKeyId,
      'blockSdoc',
      { kind: 'sdoc', sdocId: params.targetSdocId },
      issuedAt,
      { reason: params.reason }
    );
    await this.deps.revocationStore.addBlockedSdoc(params.targetSdocId, { issuedAt, reason: params.reason, statementBytes: built.bytes, byKeyId: params.signerKeyId });
    return built;
  }

  /** Reverse a block of an individual SDoc. */
  async unblockSdoc(params: BlockSdocParams): Promise<StatementResult> {
    const sdocBytes = await this.deps.documentStore.get(params.targetSdocId);
    if (!sdocBytes) throw new QrsNotFoundError(`SDoc not found: ${params.targetSdocId}`);
    const parsed = parseSignedObject(sdocBytes);
    const tcertKeyId = parsed.signerKeyId;
    const tcertId = tcertIdOf(tcertKeyId, tcertNumberOf(parsed));
    const authorized =
      params.signerKeyId === tcertKeyId || (await this.isAuthorizedCa(params.signerKeyId, tcertId));
    if (!authorized) throw new QrsAuthorizationError('Signer is not authorized to unblock this SDoc');
    const issuedAt = params.issuedAt ?? this.deps.clock.now();
    const built = await this.signStatement(
      params.signerKeyId,
      'unblockSdoc',
      { kind: 'sdoc', sdocId: params.targetSdocId },
      issuedAt,
      { reason: params.reason }
    );
    await this.deps.revocationStore.removeBlockedSdoc(params.targetSdocId, {
      issuedAt,
      reason: params.reason,
      statementBytes: built.bytes,
      byKeyId: params.signerKeyId,
    });
    return built;
  }

  /** Lifecycle check used by the verification pipeline. */
  async checkRevocation(tcertId: TcertId, keyId: KeyId, issuedAt: number, sdocId: SdocId): Promise<RevocationCheck> {
    const keyEntry = await this.deps.revocationStore.getRevokedKey(keyId);
    if (keyEntry) {
      return { state: 'invalid', message: `issuer key revoked (${keyEntry.reason ?? 'key compromised'})` };
    }
    const tcertEntry = await this.deps.revocationStore.getRevokedTcert(tcertId);
    if (tcertEntry) {
      if (tcertEntry.type === 'retrospective') {
        return { state: 'invalid', message: 'TCert retrospectively revoked' };
      }
      if (issuedAt >= tcertEntry.issuedAt) {
        return { state: 'invalid', message: 'TCert prospectively revoked for documents issued at/after revocation' };
      }
      return { state: 'valid', message: 'document issued before prospective revocation' };
    }
    const block = await this.deps.revocationStore.getBlockedSdoc(sdocId);
    if (block) return { state: 'invalid', message: 'SDoc is blocked' };
    return { state: 'valid' };
  }

  /** Is the signer an authorized CA for the target TCert? */
  async isAuthorizedCa(signerKeyId: KeyId, targetTcertId: TcertId): Promise<boolean> {
    const attestations = await this.deps.trustStore.getAttestations(targetTcertId);
    for (const att of attestations) {
      if (att.caKeyId !== signerKeyId) continue;
      const caBytes = await this.deps.certificateStore.get(att.caTcertId);
      if (!caBytes) continue;
      try {
        parseSignedObject(caBytes);
      } catch {
        continue;
      }
      if (await this.deps.revocationStore.getRevokedTcert(att.caTcertId)) continue;
      if (await this.deps.revocationStore.getRevokedKey(signerKeyId)) continue;
      if (await this.deps.trustStore.isDistrusted(att.caTcertId)) continue;
      if (!(await this.deps.trustStore.isPinned(att.caTcertId)) && !(await this.deps.trustStore.isCa(att.caTcertId))) {
        continue;
      }
      return true;
    }
    return false;
  }

  private async signStatement(
    signerKeyId: KeyId,
    action: 'revokeAttestation' | 'revokeTcert' | 'blockSdoc' | 'unblockSdoc',
    target: { kind: 'tcert'; keyId: string; certificateNumber: number } | { kind: 'key'; keyId: string } | { kind: 'sdoc'; sdocId: string },
    issuedAt: number,
    options: { reason?: string; revocationType?: RevocationType }
  ): Promise<StatementResult> {
    const priv = await this.deps.privateKeyStore.load(signerKeyId);
    if (!priv) throw new QrsNotFoundError(`Signer private key not available: ${signerKeyId}`);
    const pub = await this.deps.publicKeyStore.load(signerKeyId);
    if (!pub) throw new QrsNotFoundError(`Signer public key not found: ${signerKeyId}`);
    const provider = this.deps.cryptoRegistry.get(pub.algorithm);
    const keyPair = { algorithm: pub.algorithm, publicJwk: pub.publicJwk as PublicJwk, privateJwk: priv.privateJwk };
    const built = await buildStatement(action, target, issuedAt, options, keyPair, provider);
    if (!(await verifyStatement(built.parsed, pub.publicJwk, provider))) {
      throw new QrsCryptoError('Statement failed signature verification');
    }
    return { statementId: built.statementId, bytes: built.bytes };
  }
}
